### ISSUE TITLE
Can't use location feature of the Ar.js

### ISSUE THREAD

**Issue by knightMash** [AR-js-org/AR.js#137]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
The sphere should have shown at a location specified in the code but it is being shown at my current position only. Location feature is not working in Ar.js.

**If the current behavior is a bug, please provide the steps to reproduce.**
Here is the code 
HTML

```
<!doctype html>
<html>

<head>
    <link rel="canonical" href="https://inspiredlabs.github.io/ar.js/markerless.html" />
    <!-- location based aframe v0.9.2 -->
    <script src="https://aframe.io/releases/0.9.2/aframe.min.js"></script>
    <script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script><!-- debug -->
    <script>
    const log = console.log;
    window.onload = () => {
      let scene = document.querySelector('a-scene'); /* Apply to whole scene */

      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(function (position) {
          let gps = document.createAttribute('gps-entity-place'),
            arjs = document.createAttribute('arjs'),
            welcome = document.getElementById('welcome');

          arjs.value = 'sourceType: webcam; sourceWidth: 1280; sourceHeight: 960; trackingMethod: best; debugUIEnabled: false;';
          gps.value = `latitude: ${28.705654}; longitude: ${77.122867}`;
          log(gps.value);
          scene.setAttributeNode(gps); /* Apply to whole scene */
          scene.setAttributeNode(arjs);
        });
      }

    };
  </script>
</head>
<a-scene vr-mode-ui="enabled: false">
    <a-entity id="wrapper" position="0 -8 0">

        <a-sphere position="0 1.25 -5" radius="1.25" color="#EF2D5E" shadow></a-sphere>


        <a-text id="welcome" value="Welcome" scale="75 75 75" color="#000000" position="-30 0 -150"></a-text>
        <!-- look-at="[gps-camera]" -->

    </a-entity><!-- /wrapper -->

    <a-camera camera="fov: 60;" gps-camera rotation-reader></a-camera>
</a-scene>
</body>
</html>
```

MainActivity.java in android webview

```
public class MainActivity extends AppCompatActivity {


    private static final int REQUEST_FINE_LOCATION = 1;


    WebView webView;
    MainActivity mainActivity ;
    private Context mContext;
    
    public static final int MY_PERMISSIONS_REQUEST_ACCESS_LOCATION = 200;
    public static final int MY_PERMISSIONS_REQUEST_ACCESS_CAMERA = 1001;
    
    double bearing;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        this.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,WindowManager.LayoutParams.FLAG_FULLSCREEN);
        super.onCreate(savedInstanceState);

        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        setContentView(R.layout.activity_main);
        

        webView = (WebView)findViewById(R.id.mybrowser);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebChromeClient(new MyWebLogger());

        webView.setScrollBarStyle(WebView.SCROLLBARS_OUTSIDE_OVERLAY);
        webView.setScrollbarFadingEnabled(false);
        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setAppCacheEnabled(true);
        webView.getSettings().setDatabaseEnabled(true);
        webView.getSettings().setDomStorageEnabled(true);
        webView.getSettings().setGeolocationEnabled(true);

        webView.getSettings().setGeolocationDatabasePath(getFilesDir().getPath());
        mContext = this;
        if (Build.VERSION.SDK_INT > 17) {
            webView.getSettings().setMediaPlaybackRequiresUserGesture(false);
        }


        
        LocationManager locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);

        if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)){

            Log.i("MainActivity", "GPS is Enabled in your devide");
            if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) ==
                    PackageManager.PERMISSION_GRANTED &&
                    ContextCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_COARSE_LOCATION) ==
                            PackageManager.PERMISSION_GRANTED &&
                    ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED ) {


                webView.loadUrl("https://myurl/markerless.html");


            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, 1001);
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, 2);
            }


        }else{
            showGPSDisabledAlertToUser();

        }


        webView.setWebViewClient(new WebViewClient(){

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                Log.i("Mainactivity","Inside shouldOverride");
                webView.loadUrl(url);
                return true;
            }



        });

        webView.setWebChromeClient(new WebChromeClient() {
            public void onGeolocationPermissionsShowPrompt(String origin, GeolocationPermissions.Callback callback) {
                callback.invoke(origin, true, false);
            }
        });

        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onPermissionRequest(final PermissionRequest request) {
                final String[] requestedResources;
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    requestedResources = request.getResources();
                    for (String r : requestedResources) {
                        if (r.equals(PermissionRequest.RESOURCE_VIDEO_CAPTURE)) {
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                                request.grant(new String[]{PermissionRequest.RESOURCE_VIDEO_CAPTURE});
                            }
                            break;
                        }
                    }
                }

            }
        });


    }



    @Override
    public void onResume() {
        super.onResume();
        LocationManager locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);

        if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)){

            Log.i("MainActivity", "GPS is Enabled in your devide");
            if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) ==
                    PackageManager.PERMISSION_GRANTED &&
                    ContextCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_COARSE_LOCATION) ==
                            PackageManager.PERMISSION_GRANTED) {

                webView.loadUrl("https://myurl/markerless.html");

            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, 2);
            }

        }

    }




    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        Log.i("MainActivity", "Inside onRequestPermissionResult "+ requestCode);
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_ACCESS_LOCATION: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    // permission was granted
                    Log.i("MainActivity","User Allowed permissions");
                    webView.loadUrl("https://myurl/markerless.html");



                } else {


                    Log.i("MainActivity","User Denied Permission 0()");
                }

            }
            break;
            case MY_PERMISSIONS_REQUEST_ACCESS_CAMERA: {
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    //Do your stuff
                    Log.i("MainActivity","User Allowed permissions Camera");
                } else {

                }
            }
            break;

        }
    }




    private void showGPSDisabledAlertToUser(){
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(this);
        alertDialogBuilder.setMessage("GPS is disabled in your device. Would you like to enable it?")
                .setCancelable(false)
                .setPositiveButton("Goto Settings Page To Enable GPS",
                        new DialogInterface.OnClickListener(){
                            public void onClick(DialogInterface dialog, int id){
                                Intent callGPSSettingIntent = new Intent(
                                        android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                                startActivity(callGPSSettingIntent);
                            }
                        });
        alertDialogBuilder.setNegativeButton("Cancel",
                new DialogInterface.OnClickListener(){
                    public void onClick(DialogInterface dialog, int id){
                        dialog.cancel();
                    }
                });
        AlertDialog alert = alertDialogBuilder.create();
        alert.show();
    }

}
```

Logs in android studios 

```
07-05 13:33:46.355 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(0)] "The connection used to load resources from https://www.zamit.one used TLS 1.0 or TLS 1.1, which are deprecated and will be disabled in the future. Once disabled, users will be prevented from loading these resources. The server should enable TLS 1.2 or later. See https://www.chromestatus.com/feature/5654791610957824 for more information.", source: https://www.zamit.one/markerless.html (0)
07-05 13:33:47.176 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(140)] "No DPDB device match.", source: https://aframe.io/releases/0.9.2/aframe.min.js (140)
07-05 13:33:47.176 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(140)] "Failed to recalculate device parameters.", source: https://aframe.io/releases/0.9.2/aframe.min.js (140)
07-05 13:33:47.183 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(140)] "Using fallback Android device measurements.", source: https://aframe.io/releases/0.9.2/aframe.min.js (140)
07-05 13:33:47.399 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(259)] "document.registerElement is deprecated and will be removed in M73, around March 2019. Please use window.customElements.define instead. See https://www.chromestatus.com/features/4642138092470272 for more details.", source: https://aframe.io/releases/0.9.2/aframe.min.js (259)
07-05 13:33:47.520 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(353)] "A-Frame Version: 0.9.2 (Date 2019-05-06, Commit #f57a1fa)", source: https://aframe.io/releases/0.9.2/aframe.min.js (353)
07-05 13:33:47.520 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(353)] "three Version (https://github.com/supermedium/three.js):", source: https://aframe.io/releases/0.9.2/aframe.min.js (353)
07-05 13:33:47.521 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(353)] "WebVR Polyfill Version:", source: https://aframe.io/releases/0.9.2/aframe.min.js (353)
07-05 13:33:47.806 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(140)] "No DPDB device match.", source: https://aframe.io/releases/0.9.2/aframe.min.js (140)
07-05 13:33:47.807 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(140)] "Failed to recalculate device parameters.", source: https://aframe.io/releases/0.9.2/aframe.min.js (140)
07-05 13:33:48.429 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(95)] "THREE.WebGLRenderer", source: https://aframe.io/releases/0.9.2/aframe.min.js (95)
07-05 13:33:49.551 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(4120)] "ERROR(1): User denied Geolocation", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (4120)
07-05 13:33:53.448 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2612)] "AR.js", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2612)
07-05 13:33:54.930 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2)] "[warning] *** Camera Parameter resized from 640, 480. ***", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2)
07-05 13:33:55.161 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2)] "[info] Allocated videoFrameSize 307200", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2)
07-05 13:33:55.175 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2)] "[info] Pattern detection mode set to 1.", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2)
07-05 13:33:55.189 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2)] "[info] Pattern ratio size set to 0.500000.", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2)
07-05 13:33:55.189 32209-32209/com.example.sample I/chromium: [INFO:CONSOLE(2)] "[info] Labeling mode set to 1", source: https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js (2)
```

A pop up with this information is shown before opening of the app

![Screenshot_2020-07-03-14-51-34-152_com example ghostfinder](https://user-images.githubusercontent.com/61133432/86580588-8d9e8100-bf9c-11ea-9fc2-56263353071a.png)

But I have allowed all the permissions and the site also but still the pop up.

**Please mention other relevant information such as the browser version, Operating System and Device Name**
Browser - Chrome 83.0.4103.106
Os - Android 6.0.1; MarshMellow
Device-  Xiaomi Redmi 3S Prime

**What is the expected behavior?**
The Sphere is should have shown at the location specified.

**If this is a feature request, what is motivation or use case for changing the behavior?**

So basically does the location feature of Ar.js is in production? I mean does it work ? if yes , where did I went wrong?

**Comment by meizano** [comment]
Answer by Chris Cashwell in StackOverflow seems to answer your question:
The WebView must use a custom WebChromeClient which implements WebChromeClient.onGeolocationPermissionsShowPrompt().
Reference: [https://stackoverflow.com/questions/5329662/android-webview-geolocation](https://stackoverflow.com/questions/5329662/android-webview-geolocation)

**Comment by knightMash** [comment]
The culprit was using the setWebChromeClient() twice at different places as shown below.

 ```
  webView.setWebChromeClient(new WebChromeClient() {
        public void onGeolocationPermissionsShowPrompt(String origin, GeolocationPermissions.Callback callback) {
            callback.invoke(origin, true, false);
        }
    });

    webView.setWebChromeClient(new WebChromeClient() {
        @Override
        public void onPermissionRequest(final PermissionRequest request) {
            final String[] requestedResources;
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                requestedResources = request.getResources();
                for (String r : requestedResources) {
                    if (r.equals(PermissionRequest.RESOURCE_VIDEO_CAPTURE)) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                            request.grant(new String[]{PermissionRequest.RESOURCE_VIDEO_CAPTURE});
                        }
                        break;
                    }
                }
            }

        }
    });


}
```
changed it to a single setWebChromeClient() (like the following) and the first pop up vanished. The second pop-up only appears once now on installation of the app and does not pose any problem as such till now.(if it does in future, I will update the answer).

```
webView.setWebChromeClient(new WebChromeClient() {
            public void onGeolocationPermissionsShowPrompt(String origin, GeolocationPermissions.Callback callback) {
                callback.invoke(origin, true, false);
            }

            @Override
            public void onPermissionRequest(final PermissionRequest request) {
                final String[] requestedResources;
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                    requestedResources = request.getResources();
                    for (String r : requestedResources) {
                        if (r.equals(PermissionRequest.RESOURCE_VIDEO_CAPTURE)) {
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                                request.grant(new String[]{PermissionRequest.RESOURCE_VIDEO_CAPTURE});
                            }
                            break;
                        }
                    }
                }

            }
        });


    }
```
